File: src/main/java/com/nexusai/application/agent/lsp/MockLspServer.java
package com.nexusai.application.agent.lsp;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.util.Map;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * 真实 LSP server 测试替身 (stdio 子进程模式).
 *
 * <p>运行方式: {@code java -cp <classpath> com.nexusai.application.agent.lsp.MockLspServer},
 * 通过 stdin/stdout 与 ProcessLspClient 通信, 帧格式 Content-Length + body.
 * 在单测中可作为 ProcessBuilder.command() 的可执行命令启动.
 */
public class MockLspServer {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    public static void main(String[] args) throws IOException {
        new MockLspServer().run();
    }

    private void run() throws IOException {
        AtomicBoolean running = new AtomicBoolean(true);
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in, StandardCharsets.UTF_8));
        OutputStream out = System.out;
        Runtime.getRuntime().addShutdownHook(new Thread(() -> running.set(false)));
        while (running.get()) {
            String contentLengthLine = in.readLine();
            if (contentLengthLine == null) break;
            if (!contentLengthLine.toLowerCase().startsWith("content-length:")) continue;
            int len = Integer.parseInt(contentLengthLine.split(":")[1].trim());
            in.readLine(); // 空行
            char[] buf = new char[len];
            int read = 0;
            while (read < len) {
                int n = in.read(buf, read, len - read);
                if (n < 0) break;
                read += n;
            }
            String json = new String(buf, 0, read);
            JsonNode node;
            try {
                node = MAPPER.readTree(json);
            } catch (Exception e) {
                continue;
            }
            String method = node.path("method").asText("");
            long id = node.path("id").asLong(0);
            switch (method) {
                case "initialize" -> sendResponse(out, id, buildInitializeResult());
                case "textDocument/hover" -> sendResponse(out, id, buildHoverResult("mock hover content"));
                case "textDocument/definition" -> sendResponse(out, id, buildDefinitionResult());
                default -> {
                    // ignore
                }
            }
        }
    }

    private static ObjectNode buildInitializeResult() {
        ObjectNode result = MAPPER.createObjectNode();
        result.put("protocolVersion", "2024-11-05");
        ObjectNode caps = MAPPER.createObjectNode();
        caps.put("definitionProvider", true);
        caps.put("hoverProvider", true);
        caps.put("referencesProvider", true);
        caps.put("documentSymbolProvider", true);
        result.set("capabilities", caps);
        ObjectNode serverInfo = MAPPER.createObjectNode();
        serverInfo.put("name", "mock-lsp-server");
        serverInfo.put("version", "1.0.0-test");
        result.set("serverInfo", serverInfo);
        return result;
    }

    private static ObjectNode buildHoverResult(String content) {
        ObjectNode result = MAPPER.createObjectNode();
        ObjectNode contents = MAPPER.createObjectNode();
        contents.put("kind", "markdown");
        contents.put("value", content);
        result.set("contents", contents);
        return result;
    }

    private static ObjectNode buildDefinitionResult() {
        ObjectNode result = MAPPER.createObjectNode();
        ObjectNode location = MAPPER.createObjectNode();
        location.put("uri", "file:///mock/definition.ts");
        ObjectNode range = MAPPER.createObjectNode();
        ObjectNode start = MAPPER.createObjectNode();
        start.put("line", 0);
        start.put("character", 0);
        ObjectNode end = MAPPER.createObjectNode();
        end.put("line", 0);
        end.put("character", 10);
        range.set("start", start);
        range.set("end", end);
        location.set("range", range);
        result.set("location", location);
        return result;
    }

    private static void sendResponse(OutputStream out, long id, Object result) throws IOException {
        Map<String, Object> msg = Map.of(
            "jsonrpc", "2.0",
            "id", id,
            "result", result
        );
        byte[] body = MAPPER.writeValueAsBytes(msg);
        String header = "Content-Length: " + body.length + "\r\n\r\n";
        out.write(header.getBytes(StandardCharsets.US_ASCII));
        out.write(body);
        out.flush();
    }
}

File: src/test/resources/mock-lsp-server.py
#!/usr/bin/env python3
"""
Mock LSP server for testing ProcessLspClient · stdio JSON-RPC over LSP framed protocol.
Reads Content-Length framed JSON requests, responds to initialize/hover/definition.
Records didOpen/didChange/didSave notifications and exposes them via test/getNotifications.
Usage: launched as subprocess by ProcessLspClient tests.
"""
import sys
import json
import threading

_lock = threading.Lock()
_notifications = []


def send_response(id_, result):
    body = {"jsonrpc": "2.0", "id": id_, "result": result}
    payload = json.dumps(body).encode("utf-8")
    sys.stdout.buffer.write(f"Content-Length: {len(payload)}\r\n\r\n".encode("ascii"))
    sys.stdout.buffer.write(payload)
    sys.stdout.buffer.flush()


INIT_RESULT = {
    "protocolVersion": "2024-11-05",
    "capabilities": {
        "definitionProvider": True,
        "hoverProvider": True,
        "referencesProvider": True,
        "documentSymbolProvider": True,
    },
    "serverInfo": {"name": "mock-lsp-server", "version": "1.0.0-test"},
}
HOVER_RESULT = {"contents": {"kind": "markdown", "value": "mock hover content"}}
DEF_RESULT = {"location": {"uri": "file:///mock/definition.ts", "range": {
    "start": {"line": 0, "character": 0}, "end": {"line": 0, "character": 10}}}}

# LSP sync notifications to record (method + params) for test/getNotifications verification.
SYNC_NOTIFICATIONS = {
    "textDocument/didOpen",
    "textDocument/didChange",
    "textDocument/didSave",
}


def record_notification(method, params):
    with _lock:
        _notifications.append({"method": method, "params": params})


def main():
    stdin = sys.stdin.buffer
    while True:
        # Read headers
        content_length = 0
        while True:
            line = stdin.readline()
            if not line:
                return  # EOF
            line = line.strip()
            if not line:
                break
            if line.lower().startswith(b"content-length:"):
                content_length = int(line.split(b":", 1)[1].strip())
        if content_length <= 0:
            continue
        # Read body
        body_bytes = stdin.read(content_length)
        if not body_bytes:
            return
        try:
            msg = json.loads(body_bytes)
        except Exception:
            continue
        method = msg.get("method", "")
        id_ = msg.get("id")
        if method == "initialize":
            send_response(id_, INIT_RESULT)
        elif method == "textDocument/hover":
            send_response(id_, HOVER_RESULT)
        elif method == "textDocument/definition":
            send_response(id_, DEF_RESULT)
        elif method == "test/getNotifications":
            with _lock:
                send_response(id_, list(_notifications))
        elif id_ is None and method in SYNC_NOTIFICATIONS:
            # fire-and-forget notification → record for E2E verification
            record_notification(method, msg.get("params", {}))
        # unknown methods: silent (notification or unimplemented)


if __name__ == "__main__":
    main()
